File: December 2024 Leetcode Solution/December-08.java
class Solution {
    public int maxTwoEvents(int[][] events) {
        Arrays.sort(events, (a, b) -> Integer.compare(a[0], b[0]));

        int n = events.length;
        int[] maxValues = new int[n];
        maxValues[n - 1] = events[n - 1][2];
        for (int i = n - 2; i >= 0; --i) {
            maxValues[i] = Math.max(maxValues[i + 1], events[i][2]);
        }

        int maxResult = 0;
        for (int i = 0; i < n; ++i) {
            int nextIndex = binarySearch(events, events[i][1]);
            int currentValue = events[i][2];
            if (nextIndex < n) {
                currentValue += maxValues[nextIndex];
            }
            maxResult = Math.max(maxResult, currentValue);
        }

        return maxResult;
    }

    private int binarySearch(int[][] events, int target) {
        int low = 0, high = events.length - 1;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (events[mid][0] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
}

File: December 2024 Leetcode Solution/December-08.rs
impl Solution {
    pub fn max_two_events(events: Vec<Vec<i32>>) -> i32 {
        let mut events = events;
        events.sort();

        let n = events.len();
        let mut max_values = vec![0; n];
        max_values[n - 1] = events[n - 1][2];

        for i in (0..n - 1).rev() {
            max_values[i] = max_values[i + 1].max(events[i][2]);
        }

        let mut max_result = 0;
        for i in 0..n {
            let value = events[i][2];
            let idx = events.binary_search_by(|event| event[0].cmp(&(events[i][1] + 1))).unwrap_or_else(|x| x);
            let total_value = if idx < n { value + max_values[idx] } else { value };
            max_result = max_result.max(total_value);
        }

        max_result
    }
}
